libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/policy/FileUtils.java

```diff
@@ -44,17 +65,51 @@ private FileUtils() {}
                 if (c2IsSeparator) {
                     return 1;
                 }
-                return c1 - c2;
+                return comp;
             }
         }
         return len1 - len2;
     };
 
+    /**
+     * Case-insensitive character comparison. Inspired by the {@code WindowsPath#compareTo} implementation.
+     */
+    private static int caseInsensitiveCharacterComparator(char c1, char c2) {
+        if (c1 == c2) {
+            return 0;
+        }
+        c1 = Character.toUpperCase(c1);
+        c2 = Character.toUpperCase(c2);
+        if (c1 == c2) {
+            return 0;
+        }
+        return c1 - c2;
+    }
+
+    private static int caseSensitiveCharacterComparator(char c1, char c2) {
+        return c1 - c2;
```

Comment:
Using subtraction to implement a comparator is not recommended, because there are overflow corner cases that don't work properly. We should call `Character.compare` here.

It turns out that, for `char` in particular, it makes no difference because the overflow can't happen, but calling `Character.compare` means folks reading this code don't even need to worry about this issue at all.

Reply:
Again, copy-pasted this from WindowsPath in the JDK, so we should be fine; but I see your point of changing it for the sake of people reading the code. I can do it.